File: velolabs_repos/lattis_android/app/src/main/java/com/lattis/ellipse/domain/interactor/lock/base/BaseLockUseCase.java
package com.lattis.ellipse.domain.interactor.lock.base;

import com.lattis.ellipse.domain.executor.PostExecutionThread;
import com.lattis.ellipse.domain.executor.ThreadExecutor;
import com.lattis.ellipse.domain.interactor.UseCase;
import com.lattis.ellipse.domain.model.Lock;
import com.lattis.ellipse.domain.repository.BluetoothRepository;
import com.lattis.ellipse.domain.repository.LockRepository;

import io.reactivex.Observable;
import io.reactivex.disposables.Disposable;
import io.reactivex.observers.DisposableObserver;
import io.reactivex.schedulers.Schedulers;

/**
 * Created by ssd3 on 4/26/17.
 */

public abstract class BaseLockUseCase<ReturnType> extends UseCase<ReturnType> {

    private final BluetoothRepository bluetoothRepository;
    private final LockRepository lockRepository;
    private String lockId;
    private Lock lock;
    private final String TAG = BaseLockUseCase.class.getName();

    protected BaseLockUseCase(ThreadExecutor threadExecutor,
                              PostExecutionThread postExecutionThread,
                              BluetoothRepository bluetoothRepository,
                              LockRepository lockRepository
                              ) {
        super(threadExecutor, postExecutionThread);
        this.bluetoothRepository = bluetoothRepository;
        this.lockRepository=lockRepository;
    }


    @Override
    public Disposable execute(DisposableObserver<ReturnType> disposableObserver) {
        return this.buildUseCaseObservable()
                .subscribeOn(Schedulers.from(threadExecutor))
                .observeOn(postExecutionThread.getScheduler(), true)
                .subscribeWith(disposableObserver);
    }

    @Override
    public Disposable executeInMainThread(DisposableObserver<ReturnType> disposableObserver) {
        return this.buildUseCaseObservable()
                .observeOn(postExecutionThread.getScheduler(), true)
                .subscribeWith(disposableObserver);
    }

    protected BluetoothRepository getBluetoothRepository() {
        return bluetoothRepository;
    }

    protected void setScannedLock(Lock lock) {
        this.lock = lock;
    }



    public Observable<Lock.Connection.Status> connectToLock(){
        this.lockId = lock.getLockId();
        return bluetoothRepository.connectTo(lock);
    }


    public Observable<Lock.Connection.Status> connectTo(Lock lock){
        this.lock = lock;
        return bluetoothRepository.connectTo(lock);
    }

    protected Observable<Boolean> setPosition(boolean locked) {
        //return lockRepository.getLock().flatMap(connectedLock -> bluetoothRepository.setPosition(connectedLock, locked));
        return bluetoothRepository.setPosition(lock, locked);
    }

    protected Observable<Lock.Hardware.Position> getLockPositionObservable() {
        return bluetoothRepository.observePosition(lock);
    }

    protected Observable<Lock.Connection.Status> getConnectionObservable() {
        return bluetoothRepository.observeLockConnectionState((lock));
    }

    protected Observable<Lock.Hardware.State> getHardwareStateObservable() {
        return bluetoothRepository.observeHardwareState(lock);
    }


    protected Observable<Lock.Connection.Status> connectToLastConnectedLock(){
        return bluetoothRepository.getLastConnectedLock()
//                .onErrorResumeNext(throwable -> lockRepository.getLock())
                .flatMap(this::connectTo);
    }

    protected Observable<Boolean> disconnectAllLocks(){
        return bluetoothRepository.disconnectAllLocks();
    }

    protected  Observable<String> getLockFirmwareVersion(){
        return bluetoothRepository.getLockFirmwareVersion(lock);
    }

    protected  Observable<Boolean> isLockConnected(Lock lock){
        return bluetoothRepository.isConnectedTo(lock);
    }


}

File: velolabs_repos/lattis_android_v2/domain/src/main/java/com/lattis/domain/usecase/lock/base/BaseLockUseCase.kt
package com.lattis.domain.usecase.lock.base

import android.util.Log
import com.lattis.domain.executor.PostExecutionThread
import com.lattis.domain.executor.ThreadExecutor
import com.lattis.domain.models.Lock
import com.lattis.domain.repository.BluetoothRepository
import com.lattis.domain.repository.LockRepository
import com.lattis.domain.repository.SaSOrPSLockRepository
import com.lattis.domain.usecase.base.UseCase
import io.reactivex.rxjava3.core.Observable
import io.reactivex.rxjava3.disposables.Disposable
import io.reactivex.rxjava3.observers.DisposableObserver
import io.reactivex.rxjava3.schedulers.Schedulers

abstract class BaseLockUseCase<ReturnType> protected constructor(
    threadExecutor: ThreadExecutor?,
    postExecutionThread: PostExecutionThread?,
    protected val bluetoothRepository: BluetoothRepository,
    private val lockRepository: LockRepository,
    private val saSOrPSLockRepository: SaSOrPSLockRepository
) : UseCase<ReturnType>(threadExecutor!!, postExecutionThread!!) {
    private var lock: Lock? = null

    private val TAG = BaseLockUseCase::class.java.name
    override fun execute(disposableObserver: DisposableObserver<ReturnType>): Disposable {
        return buildUseCaseObservable()
            .subscribeOn(Schedulers.from(threadExecutor))
            .observeOn(postExecutionThread.scheduler, true)
            .subscribeWith(disposableObserver)
    }

    override fun executeInMainThread(disposableObserver: DisposableObserver<ReturnType>): Disposable {
        return buildUseCaseObservable()
            .observeOn(postExecutionThread.scheduler, true)
            .subscribeWith(disposableObserver)
    }

    protected fun setScannedLock(lock: Lock) {
        this.lock = lock
    }

    fun connectToLock(vendor: LockVendor): Observable<Lock.Connection.Status> {
        return when (vendor) {
            LockVendor.ELLIPSE -> {
                 bluetoothRepository.connectTo(lock!!)
            }

            LockVendor.AXA -> {
                bluetoothRepository.connectToAxa(lock!!)
            }

            LockVendor.TAPKEY -> {
                bluetoothRepository.connectToTapkey(lock!!)
            }
            LockVendor.PSLOCK, LockVendor.SAS->{
                bluetoothRepository.connectToSaSOrPSLock(lock!!,vendor)
            }
        }
    }

    fun connectTo(vendor: LockVendor, lock: Lock): Observable<Lock.Connection.Status> {
        this.lock = lock
        return connectToLock(vendor)
    }

    protected fun setPosition(vendor: LockVendor,locked: Boolean,fleetId:Int?): Observable<Boolean> { //return lockRepository.getLock().flatMap(connectedLock -> bluetoothRepository.setPosition(connectedLock, locked));
        return when(vendor){
            LockVendor.ELLIPSE ->{
                bluetoothRepository.setPosition(lock!!, locked)
            }

            LockVendor.AXA->{
                bluetoothRepository.setPositionForAxa(lock!!,locked)
            }

            LockVendor.TAPKEY->{
                bluetoothRepository.setPositionForTapkey(lock!!,locked)
            }
            LockVendor.PSLOCK, LockVendor.SAS->{
                 bluetoothRepository.getNonceTokenForSasOrPSLock(lock!!,vendor).flatMap { nonce ->
                     val device_id = if(lock?.macId!=null)lock?.macId else if(lock?.macAddress!=null) lock?.macAddress else null
//                     val device_id = "50B0456022"
//                     val device_id = "60DD985F96"
                     if(device_id!=null && fleetId!=null) {

                         return@flatMap saSOrPSLockRepository.getUnlockToken(fleetId,device_id!!, nonce)
                             .flatMap {
                                 lock?.signedMessage = it.saSOrPSLockUnlockToken?.token
                                 bluetoothRepository.setPositionForSaSOrPSLock(
                                     lock!!,
                                     locked,
                                     vendor
                                 )
                             }
                     }else{
                         return@flatMap Observable.just(false)
                     }
                 }
            }
        }
    }

    protected fun lockPositionObservable(vendor: LockVendor): Observable<Lock.Hardware.Position> {

        return when(vendor){
            LockVendor.ELLIPSE ->{
                bluetoothRepository.observePosition(lock!!)
            }

            LockVendor.AXA->{
                bluetoothRepository.observeAxaPosition(lock!!)
            }

            LockVendor.TAPKEY->{
                bluetoothRepository.observeTapkeyPosition(lock!!)
            }
            LockVendor.PSLOCK, LockVendor.SAS->{
                bluetoothRepository.observeSaSOrPSLockPosition(lock!!,vendor)
            }
        }

    }

    protected val connectionObservable: Observable<Lock.Connection.Status>
        protected get() = bluetoothRepository.observeLockConnectionState(lock!!)

    protected fun hardwareStateObservable(vendor: LockVendor): Observable<Lock.Hardware.State>{
        return when(vendor){
            LockVendor.ELLIPSE ->{
                bluetoothRepository.observeHardwareState(lock!!)
            }

            LockVendor.AXA->{
                bluetoothRepository.observeAxaHardwareState(lock!!)
            }

            LockVendor.TAPKEY->{
                bluetoothRepository.observeTapkeyHardwareState(lock!!)
            }
            LockVendor.PSLOCK, LockVendor.SAS->{
                bluetoothRepository.observeSaSOrPSLockHardwareState(lock!!,vendor)
            }
        }
    }

    protected fun connectToLastConnectedLock(vendor: LockVendor): Observable<Lock.Connection.Status> {
        return when(vendor){
            LockVendor.ELLIPSE ->{
                bluetoothRepository.getLastConnectedLock()
            }

            LockVendor.AXA->{
                bluetoothRepository.getLastConnectedAxaLock()
            }

            LockVendor.TAPKEY->{
                bluetoothRepository.getLastConnectedTapkeyLock()
            }

            LockVendor.PSLOCK, LockVendor.SAS->{
                bluetoothRepository.getLastConnectedSaSOrPSLock(vendor)
            }
        }.flatMap { lock: Lock -> connectTo(vendor,lock!!) }
    }

    protected fun disconnectAllLocks(vendor: LockVendor): Observable<Boolean> {
        return return when(vendor){
            LockVendor.ELLIPSE ->{
                bluetoothRepository.disconnectAllLocks()
            }

            LockVendor.AXA->{
                bluetoothRepository.disconnectAllAxaLocks()
            }

            LockVendor.TAPKEY->{
                bluetoothRepository.disconnectAllTapkeyLocks()
            }
            LockVendor.PSLOCK, LockVendor.SAS->{
                bluetoothRepository.disconnectAllSaSOrPSLocks(vendor)
            }
        }
    }

    protected val lockFirmwareVersion: Observable<String>
        protected get() = bluetoothRepository.getLockFirmwareVersion(lock!!)

    protected fun isLockConnected(lock: Lock): Observable<Boolean> {
        return bluetoothRepository.isConnectedTo(lock!!)
    }

}